rosbridge_library/src/rosbridge_library/internal/message_conversion.py

```diff
@@ -207,71 +229,78 @@ def msg_class_type_repr(msg_class):
     return f"{class_repr[0]}/{class_repr[1]}/{class_repr[3]}"
 
 
-def _from_inst(inst, rostype):
-    global bson_only_mode
+def _from_inst(
+    inst: ROSMessage | ListType | PrimitiveType | bytes, rostype: str
+) -> dict | list | PrimitiveType | bytes | None:
     # Special case for uint8[], we encode the string
     for binary_type, expression in ros_binary_types_list_braces:
         if expression.sub(binary_type, rostype) in ros_binary_types:
+            assert isinstance(inst, list_types)
             encoded = get_encoder()(inst)
             return encoded.decode("ascii")
 
     # Check for time or duration
     if rostype in ros_time_types:
+        assert isinstance(inst, TimeMsg | DurationMsg)
         return {"sec": inst.sec, "nanosec": inst.nanosec}
 
-    if bson_only_mode is None:
-        bson_only_mode = rospy.get_param("~bson_only_mode", False)
     # Check for primitive types
     if rostype in ros_primitive_types:
-        # JSON does not support Inf and NaN. They are mapped to None and encoded as null
-        if (
-            not bson_only_mode
-            and rostype in type_map.get("float")
-            and (math.isnan(inst) or math.isinf(inst))
-        ):
-            return None
-
-        # JSON does not support byte array. They are converted to int
-        if (not bson_only_mode) and (rostype == "octet"):
-            return int.from_bytes(inst, "little")
-
-        return inst
+        assert isinstance(inst, (*primitive_types, bytes))
+        return _from_primitive_inst(inst, rostype)
 
     # Check if it's a list or tuple
-    if type(inst) in list_types:
+    if isinstance(inst, list_types):
         return _from_list_inst(inst, rostype)
 
     # Assume it's otherwise a full ros msg object
+    assert isinstance(inst, ROSMessage)
     return _from_object_inst(inst, rostype)
 
 
-def _from_list_inst(inst, rostype):
+def _from_primitive_inst(inst: PrimitiveType | bytes, rostype: str) -> PrimitiveType | bytes | None:
+    if bson_only_mode:
+        return inst
+
+    # JSON does not support Inf and NaN. They are mapped to None and encoded as null
+    if rostype in type_map["float"]:
+        assert isinstance(inst, float)
+        if math.isnan(inst) or math.isinf(inst):
+            return None
+
+    # octet is translated to byte array with length 1
+    # JSON does not support byte array. They are converted to int
+    if rostype == "octet":
+        assert isinstance(inst, bytes)
```

Comment:
I think asserts are not favorable, as they will just raise an `AssertionError` instead of manually raising a `TypeError`.